Support StrictRedis.eval for Lua scripts

blfoster opened the pull request:
Fixes https://github.com/jamesls/fakeredis/issues/176

blfoster committed:
Limited Lua support

blfoster committed:
Bump patch version

blfoster committed:
Remove nightly Python build

blfoster committed:
Return result from Lua

blfoster committed:
Fix table handling

blfoster committed:
Fix unit tests

blfoster committed:
Fix test

blfoster committed:
Decode non-byte strings

blfoster committed:
Add eval

blfoster committed:
Restore nightly build

blfoster committed:
Revert "Restore nightly build"

This reverts commit 06e02b5d40e7b17ae69a4d850c3b64aab5b1144b.

blfoster committed:
Update readme

blfoster committed:
Remove eval from unimplemented functions

blfoster commented:
Coveralls report shows that new code has 100% coverage. I'm not sure why it reports a net decrease. The uncovered line seems unrelated to this PR.

bmerry commented:
Thanks, I'll take a look soon. What happens if the user doesn't have Lua installed? Ideally it should gracefully fall back to not supporting the `eval` command, rather than being a hard requirement.

blfoster commented:
Thanks. I'm testing it now on a clean install of Linux Mint without Lua, and it works fine. I don't have a great explanation for how that is possible at the moment. I would need to dig into the lupa source a bit more. I can look at that in the morning.

bmerry commented:
Is it possible that lupa is bundling a local copy of Lua in a binary wheel? That might be something that will work only for people using platforms for which lupa ships a wheel.

bmerry commented on fakenewsredis.py:
```diff
@@ -701,6 +704,55 @@ def sort(self, name, start=None, num=None, by=None, get=None, desc=False,
+            'incrby': self.incr
+        }
+        op = op.lower()
+        func = special_cases[op] if op in special_cases else getattr(self, op)
```
There are presumably attributes of `self` that shouldn't be visible to Lua - for a start, anything with an underscore prefix, plus things like `delete` and `from_url`. Also, when `self` is a FakeRedis rather than a FakeStrictRedis, it presumably should use the overload from the FakeStrictRedis base class.

blfoster replied:
Probably yes. This is a bit of a hack, but I figure the tradeoff is that it doesn't have a huge mapping from Redis commands to Python functions that needs to be maintained as more Redis functions get added. I'll make it a bit more selective.

blfoster commented:
It looks like Lupa includes the Lua source, which it builds as a fallback in the event that it can't find Lua installed locally.

bmerry commented:
> It looks like Lupa includes the Lua source, which it builds as a fallback in the event that it can't find Lua installed locally.

I'd prefer not to make a compiler (or an already-installed Lua) a hard requirement. I would suggest
- importing lupa locally in the eval function; and
- declaring it as an `extras_require` in setup.py.

blfoster committed:
Code review fallout

blfoster committed:
Code review fallout

blfoster committed:
pep8:

blfoster committed:
pep8:

blfoster committed:
Install lupa for CI

blfoster committed:
Install lupa for CI

blfoster commented:
Code review fallout is complete.

blfoster committed:
Merge branch 'master' of ../fakeredis

bmerry commented on README.rst:
```diff
@@ -230,6 +230,7 @@ db=10 in order to minimize collisions with an existing redis instance.
 To run all the tests, install the requirements file::
 
     pip install -r requirements.txt
+    pip install .[lua]
```
Remove this, and keep `lupa` in requirements.txt. The requirements.txt is used only for deployment on Travis.

blfoster replied:
Done.

bmerry commented on setup.py:
```diff
@@ -30,5 +30,8 @@
         'redis',
-    ]
+    ],
+    extras_require=dict(
```
I prefer dict literal syntax `{'lua': ...}`.

blfoster replied:
Done.

bmerry commented on setup.py:
```diff
@@ -30,5 +30,8 @@
-    ]
+    ],
+    extras_require=dict(
+        lua=['lupa==1.6']
```
Is there a reason to require exactly this version in the abstract requirement (see [here](https://caremad.io/posts/2013/07/setup-vs-requirement/) for what I meant by abstract)?

blfoster replied:
Just copy-pasta. Fixed.

bmerry commented on fakenewsredis.py:
```diff
@@ -701,6 +702,78 @@ def sort(self, name, start=None, num=None, by=None, get=None, desc=False,
+        raw_lua = """
+        function(KEYS, ARGV, callback)
+            redis = {{}}
+            redis.call = callback
```
Also need `redis.pcall`, `redis.error_reply`, `redis.status_reply`?

blfoster replied:
The latter two are done.

bmerry reviewed (changes requested):
There is a fair amount still to do in faking the lua <-> redis <-> Python type conversion path, which is different from the lupa-provided lua <-> Python path. See https://redis.io/commands/eval. Here are a few tests I wrote that fail (but pass on TestRealStrictRedis). They only test the Lua -> Redis conversions, and don't cover all the cases.
```python
    def test_eval_syntax_error(self):
        with self.assertRaises(ResponseError):
            self.redis.eval('return "', 0)

    def test_eval_global_variable(self):
        # Redis doesn't allow script to define global variables
        with self.assertRaises(ResponseError):
            self.redis.eval('a=10', 0)

    def test_eval_convert_number(self):
        # Redis forces all Lua numbers to integer
        val = self.redis.eval('return 3.2', 0)
        self.assertEqual(val, 3)
        val = self.redis.eval('return 3.8', 0)
        self.assertEqual(val, 3)
        val = self.redis.eval('return -3.8', 0)
        self.assertEqual(val, -3)

    def test_eval_convert_bool(self):
        # Redis converts true to 1 and false to nil (which redis-py converts to None)
        val = self.redis.eval('return false', 0)
        self.assertIsNone(val)
        val = self.redis.eval('return true', 0)
        self.assertEqual(val, 1)
        self.assertNotIsInstance(val, bool)

    def test_eval_return_error(self):
        with self.assertRaises(redis.ResponseError) as cm:
            self.redis.eval('return {err="Testing"}', 0)
        self.assertIn('Testing', str(cm.exception))
        with self.assertRaises(redis.ResponseError) as cm:
            self.redis.eval('return redis.error_reply("Testing")', 0)
        self.assertIn('Testing', str(cm.exception))

    def test_eval_return_ok(self):
        val = self.redis.eval('return {ok="Testing"}', 0)
        self.assertEqual(val, b'Testing')
        val = self.redis.eval('return redis.status_reply("Testing")', 0)
        self.assertEqual(val, b'Testing')
```

bmerry commented:
Thanks very much for taking this on by the way. It's a feature I've been wanting to use in (real) redis but haven't been able to without a good way to unit test it.

There was a previous pull request to add Lua support: jamesls/fakeredis#39. I didn't merge it into fakenewsredis when I made the fork because it was several years behind and had lots of merge conflicts, but you may be able to get some pieces and tricks from there. From a quick look it seems like it probably handles a number of the conversion issues I raised, plus others I hadn't even thought of yet, and also emulates the higher-level script support provided by redis-py.

blfoster committed:
More fallout

blfoster committed:
More fallout

blfoster committed:
Remove lua from .travis.yml

blfoster committed:
Remove lua from .travis.yml

blfoster commented:
Your new tests pass now. It also works recursively (so nested tables get handled properly). I haven't gone through the docs to check for other cases I"m missing yet. I still need to add pcall.

blfoster committed:
More tests

blfoster committed:
More tests

blfoster committed:
pcall

blfoster committed:
pcall

blfoster committed:
Test nested ok response

blfoster committed:
Test nested ok response

blfoster commented:
redis.pcall works now, and I've fixed a few more cases that I found (and added tests for them). I still need to go through the docs to see if I missed anything.

blfoster committed:
pep8

blfoster committed:
pep8

blfoster committed:
Fix broken tst

blfoster committed:
Merge branch 'master' of ../fakeredis

blfoster committed:
Remove extra braces

blfoster committed:
Merge branch 'master' of ../fakeredis

blfoster committed:
Fix type

blfoster committed:
Fix decoding

blfoster committed:
Merge branch 'master' of ../fakeredis

blfoster committed:
Mess with string types some more

blfoster committed:
Merge branch 'master' of ../fakeredis

blfoster committed:
Fix test

blfoster committed:
Merge branch 'master' of ../fakeredis

bmerry commented:
>I still need to go through the docs to see if I missed anything.

Okay, let me know when you think it's all covered and I'll take another look.

blfoster committed:
Test for nil in table

blfoster committed:
More tests

blfoster committed:
Fix test

blfoster commented:
@bmerry I've added a bunch more tests and fixed some more issues. It seems that you're more familiar with Redis than I, so you may find that I've missed something, but it looks pretty good to me.

bmerry commented:
> I've added a bunch more tests and fixed some more issues. It seems that you're more familiar with Redis than I, so you may find that I've missed something, but it looks pretty good to me.

Great. I'll look over it again next week. Coveralls seems to indicate that there is an exception handler that isn't being covered by the tests - can you take a look?

blfoster committed:
Remove unnecessary exception handler

blfoster commented:
Oops, I don't think that exception handler is necessary. I took it out.

bmerry commented on README.rst:
```diff
@@ -266,6 +265,11 @@ they have all been tagged as 'slow' so you can skip them by running::
 Revision history
 ================
 
+0.9.5
```
Can you change this to "Development version". There is some possibility of re-merging with fakeredis, so I don't want to make assumptions about version numbers yet.

blfoster replied:
Done.

bmerry commented on README.rst:
```diff
@@ -266,6 +265,11 @@ they have all been tagged as 'slow' so you can skip them by running::
+0.9.5
+-----
+Add support for StrictRedis.eval for Lua scripts
+- `#9 <https://github.com/ska-sa/fakenewsredis/pull/9>`_ 
```
Can you put the description into the bullet point (to make the format of the other changelog entries).

blfoster replied:
Done.

bmerry commented on fakenewsredis.py:
```diff
@@ -701,6 +702,138 @@ def sort(self, name, start=None, num=None, by=None, get=None, desc=False,
+        Returns the result of the script.
+        In practice, use the object returned by ``register_script``. This
+        function exists purely for Redis API completion.
+        """
```
You can delete the docstring. It's assumed that people will refer to redis-py for docs.

blfoster replied:
Done.

bmerry commented on fakenewsredis.py:
```diff
@@ -701,6 +702,138 @@ def sort(self, name, start=None, num=None, by=None, get=None, desc=False,
+        expected_globals = set()
+        set_globals(
+            (None,) + keys_and_args[:numkeys],
+            (None,) + keys_and_args[numkeys:],
```
I imagine this will go wrong if numkeys is out of range; it should probably raise a ResponseError.

blfoster replied:
Indeed.

blfoster replied:
Fixed, along with a bunch of other edge cases.

bmerry commented on fakenewsredis.py:
```diff
@@ -701,6 +702,138 @@ def sort(self, name, start=None, num=None, by=None, get=None, desc=False,
+                "Script attempted to set a global variables: %s" % ", ".join(
+                    actual_globals - expected_globals
+                )
+            )
```
I don't think this is a good approach to preventing global variables being created: it doesn't make the script itself error out, and it doesn't actually prevent the globals from polluting the namespace.

[Here](https://github.com/antirez/redis/blob/32ac4c64baf00747da1acc0cc61ee236922e2dcf/src/scripting.c#L867-L885) is the code in redis itself that implements the protection. It may need some tweaks to adapt it - unfortunately I've never programmed in Lua so I'm not sure.

See also the function above the code in that link - it appears to disable `readfile` and `dofile`, presumably for security reasons.

blfoster replied:
I think this is actually safe, because if we try to set a global variable and error out, the LuaRuntime instance will never be used again; if we try to run another Lua script, we'll get a new LuaRuntime, which will not have that global variable set. Unless there's something I'm missing. I do think this way is a bit more readable, but perhaps that's because I don't know Lua. I could add a few more lines to the unit test asserting that trying to set a global variable can't have a side effect...

bmerry replied:
I hadn't spotted that the runtime gets discarded each time, so it's safer than I thought. I think it will still be different to real redis if a script sets a global variable, then modifies the database: in real redis it would error out as soon as it tries to set the global, whereas in this implementation it will get to modify the database before the error. Another case that will probably behave differently is a script that creates a global and then deletes it again, before your check.

We're starting to get to the point of diminishing returns. If you've got the time and energy to test it and fix things up, then go for it, but I realise that I've made you do a lot more work than you probably expected when you started. If you're running out of steam then this is something that can be left for another day.

blfoster replied:
:+1: I don't think it should be possible to modify the database either after setting a global, because _lua_redis_call calls _check_for_lua_globals before executing any command that could change the database.

bmerry replied:
Aha, I'd missed that subtlety too. So then probably the only case that'll behave differently is if the script creates a global and deletes it again.

blfoster replied:
Yep. I'll leave that alone for now if you're ok with it.

bmerry commented on fakenewsredis.py:
```diff
@@ -701,6 +702,138 @@ def sort(self, name, start=None, num=None, by=None, get=None, desc=False,
+                item = result[index]
+                result_list.append(self._convert_lua_result(item))
+            return result_list
+        elif isinstance(result, type(u'')):
```
Use `text_type` rather than `type(u'')`.

blfoster replied:
Done.

bmerry commented on fakenewsredis.py:
```diff
@@ -701,6 +702,138 @@ def sort(self, name, start=None, num=None, by=None, get=None, desc=False,
+        elif isinstance(result, float):
+            return int(result)
+        elif isinstance(result, bool):
+            return result and 1 or None
```
Use the newer (Python 2.5+, so actually 11 years old) conditional syntax `1 if result else None`.

blfoster replied:
Done.

bmerry commented on setup.py:
```diff
@@ -5,7 +5,7 @@
 setup(
     name='fakenewsredis',
-    version='0.9.4',
+    version='0.9.5',
```
Don't update the version yet; I'll do it when I make a new release.

blfoster replied:
Done.

bmerry commented on test_fakenewsredis.py:
```diff
@@ -3046,6 +3046,212 @@ def test_set_existing_key_persists(self):
+    def test_eval_mget_none(self):
+        self.redis.set('{foo}1', None)
+        self.redis.set('{foo}2', None)
+        val = self.redis.eval('return redis.call("mget", "{foo}1", "{foo}2")', 2, 'foo1', 'foo2')
```
This call doesn't look valid: the keys passed to eval should match the keys used in the call (also same problem in a few other tests). Also, any particular reason for including curly brackets in the key names?

blfoster replied:
Oops, I'll fix the keys. The curly braces are there because mget requires that all keys shard to the same node (which the curly braces guarantee).

bmerry commented on test_fakenewsredis.py:
```diff
@@ -3046,6 +3046,212 @@ def test_set_existing_key_persists(self):
+            [
+                val[:2],
+                val[2:]
+            ]
```
This whole statement can go on one line.

bmerry commented on test_fakenewsredis.py:
```diff
@@ -3046,6 +3046,212 @@ def test_set_existing_key_persists(self):
+
+    def test_eval_convert_nil_to_false(self):
+        val = self.redis.eval('return ARGV[1] == false', 0, None)
+        self.assertFalse(val)
```
I don't think this test is testing what is implied by the name. redis-py converts all arguments through its to_bytes, so ``None`` as an argument becomes the string "None" in the script, rather than false. The following test currently fails (but passes for real redis):
```python
        val = self.redis.eval('return ARGV[1] == "None"', 0, None)
        self.assertTrue(val)
```

blfoster replied:
Huh, something went wrong here. I'll sort it out...

blfoster replied:
Actually Lua scripts require that too, now that I think about it. I guess it doesn't matter for the real Redis tests since they just run against a single node.

blfoster replied:
Braces are gone.

blfoster replied:
Fixed.

bmerry commented on fakenewsredis.py:
```diff
@@ -701,6 +702,138 @@ def sort(self, name, start=None, num=None, by=None, get=None, desc=False,
+        try:
+            result = lua_runtime.execute(script)
+        except LuaSyntaxError as ex:
+            raise ResponseError(ex)
```
I think you need to catch more than just LuaSyntaxError. The following test fails with a LuaError.
```python
    def test_eval_runtime_error(self):
        with self.assertRaises(ResponseError):
            self.redis.eval('error("CRASH")', 0)
```

blfoster replied:
Fixed.

bmerry commented on fakenewsredis.py:
```diff
@@ -701,6 +702,138 @@ def sort(self, name, start=None, num=None, by=None, get=None, desc=False,
+            'ZREVRANGEBYLEX', 'ZREVRANGEBYSCORE', 'ZREVRANK', 'ZSCAN', 'ZSCORE', 'ZUNIONSTORE'
+        ]
+        if op.upper() not in commands:
+            raise ResponseError("Unknown Redis command called from Lua script")
```
It seems a bit odd to convert to uppercase for this check and then to lowercase to actually make the call. How about just expressing the whitelist in lowercase?

blfoster replied:
Heh, the commands are uppercase because that's the way they were written in the Redis documentation I copied this from. I'll change it to lowercase.

blfoster replied:
Done.

bmerry commented on fakenewsredis.py:
```diff
@@ -701,6 +702,138 @@ def sort(self, name, start=None, num=None, by=None, get=None, desc=False,
+        except Exception as ex:
+            return lua_runtime.table_from(
+                {"err": str(ex)}
+            )
```
This can go on one line.

blfoster replied:
Done.

blfoster committed:
WIP: fallout

blfoster committed:
Remove docstring

blfoster committed:
More fallout

blfoster committed:
Catch LuaError

blfoster committed:
Check number of keys

blfoster committed:
More fallout

blfoster commented:
There are some more subtleties to the numkeys argument (in addition to not being negative), which I've added. I'll push those in a few minutes when I have tests.

blfoster committed:
More edge cases for numkeys

blfoster committed:
Remove braces

blfoster committed:
flake8

blfoster committed:
to_bytes

blfoster commented:
Ready for review again.

bmerry commented:
I'm going to merge this because I think it's at a point where it is a solid useful contribution - thanks! There are still some other things that could be done if you're feeling keen to work on them in a new PR. If not I can file them as bugs to be worked on later:
- The other script related commands (EVALSHA, SCRIPT LOAD etc)
- The Script convenience class provided by redis-py on top of the raw commands
- Loading modules like bitop, cjson etc
- Some environmental differences e.g. corner cases in global variable handling we've discussed, blocking database modifications after calling non-deterministic functions, removal of dofile and readfile.  This is probably a rabbit-hole though.
